source/common/router/header_parser.cc

```diff
@@ -269,6 +270,18 @@ HeaderParser::configure(const Protobuf::RepeatedPtrField<HeaderValueOption>& hea
       append_action = header_value_option.append_action();
     }
 
+    // There should be only one OVERWRITE_IF_EXISTS_OR_ADD per each header.
+    if (append_action == HeaderValueOption::OVERWRITE_IF_EXISTS_OR_ADD) {
+      if (headers_to_overwrite.find(header_value_option.header().key()) !=
+          headers_to_overwrite.end()) {
+        throw EnvoyException(
+            fmt::format("Multiple OVERWRITE_IF_EXISTS_OR_ADD actions for header {} are not allowed",
+                        header_value_option.header().key()));
+      } else {
+        headers_to_overwrite.emplace(header_value_option.header().key());
```

Comment:
Should this also take into account headers inserted with `APPEND_IF_EXISTS_OR_ADD`?

Reply:
As per above comment, I suggest that we drop that check and allow not-optimal config. It does not really affect the main issue this PR addresses: circular input which grows each time it is executed.